simplex/block_builder_test.go

```diff
@@ -0,0 +1,218 @@
+
+			block, built := bb.BuildBlock(timeoutCtx, child.BlockHeader().ProtocolMetadata)
+			require.Equal(t, tt.shouldBuild, built)
+			require.Equal(t, tt.expectedBlock, block)
+			if tt.expectedBlock == nil {
+				require.Greater(t, count, 1)
+			}
+		})
+	}
+}
+
+func TestBlockBuilderCancelContext(t *testing.T) {
+	ctx := context.Background()
+	vm := newTestVM()
+	genesis := newTestBlock(t, newBlockConfig{})
+	child := newTestBlock(t, newBlockConfig{
+		prev: genesis,
+	})
+	vm.WaitForEventF = func(ctx context.Context) (common.Message, error) {
+		<-ctx.Done()
+		return 0, ctx.Err()
+	}
+
+	bb := &BlockBuilder{
+		log:          logging.NoLog{},
+		vm:           vm,
+		blockTracker: genesis.blockTracker,
+	}
+
+	timeoutCtx, cancelCtx := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancelCtx()
+
+	_, built := bb.BuildBlock(timeoutCtx, child.BlockHeader().ProtocolMetadata)
+	require.False(t, built, "Block should not be built when context is cancelled")
+}
+
+func TestWaitForPendingBlock(t *testing.T) {
+	ctx := context.Background()
+	vm := newTestVM()
+	genesis := newTestBlock(t, newBlockConfig{})
+	count := 0
+	vm.WaitForEventF = func(_ context.Context) (common.Message, error) {
+		if count == 0 {
+			count++
+			return common.StateSyncDone, nil
+		}
+		return common.PendingTxs, nil
+	}
+
+	bb := &BlockBuilder{
+		log:          logging.NoLog{},
+		vm:           vm,
+		blockTracker: genesis.blockTracker,
+	}
+
+	bb.WaitForPendingBlock(ctx)
+	require.Equal(t, 1, count)
+}
+
+func TestBlockBuildingExponentialBackoff(t *testing.T) {
+	ctx := context.Background()
+	vm := newTestVM()
+	genesis := newTestBlock(t, newBlockConfig{})
+	child := newTestBlock(t, newBlockConfig{
+		prev: genesis,
+	})
+
+	count := 0
+	vm.WaitForEventF = func(_ context.Context) (common.Message, error) {
+		count++
+		return common.PendingTxs, nil
+	}
+
+	vm.BuildBlockF = func(_ context.Context) (snowman.Block, error) {
+		if count > 7 {
+			return child.vmBlock, nil
+		}
+		return nil, errors.New("failed to build block")
+	}
+
+	bb := &BlockBuilder{
+		log:          logging.NoLog{},
+		vm:           vm,
+		blockTracker: genesis.blockTracker,
+	}
+
+	timeoutTime := 10 * time.Second
+	timeoutCtx, cancelCtx := context.WithTimeout(ctx, timeoutTime)
+	defer cancelCtx()
+
+	start := time.Now()
+	block, built := bb.BuildBlock(timeoutCtx, child.BlockHeader().ProtocolMetadata)
+	endTime := time.Since(start)
+
+	require.True(t, built)
+	require.Equal(t, child.BlockHeader(), block.BlockHeader())
+
+	// 10, 20, 40, 80, 160, 320, 640 = 1270ms
+	require.GreaterOrEqual(t, endTime.Milliseconds(), int64(1270))
+	// ensure we haven't timed out
+	require.NotEqual(t, timeoutTime, endTime)
+}
+
+func TestWaitForPendingBlockBackoff(t *testing.T) {
+	ctx := context.Background()
+	vm := newTestVM()
+
+	count := 0
+	vm.WaitForEventF = func(_ context.Context) (common.Message, error) {
+		if count < 7 {
+			count++
+			return common.StateSyncDone, nil
+		}
+		return common.PendingTxs, nil
```

Comment:
This is definitely being a bit pedantic, but this `WaitForEvent` count implementation is very similar (but different) to the `BuildBlock` count implementation above. Can we make it so that they follow the same logic? (increment the same way, ect). Bonus points if we can actually share the code in a reasonable way... But I had to understand twice why this is expected to call the functions 8 times